### ISSUE TITLE
Framework support for Ember.JS

### ISSUE THREAD

**Issue by NullVoxPopuli** [stackblitz/webcontainer-core#354]
**Describe the solution you'd like:**
I'd like stackblitz framework/template links to link to the official "output repo" for online editors (there is now an official branch just for stackblitz-specific customizations)

**Provide a link to a github repo containing an example of the framework:**

~~https://stackblitz.com/github/ember-cli/ember-new-output/tree/stable?file=README.md~~
https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output?file=README.md

This link is updated automatically with every release.

**Additional context:**

We have an influx of a need to create reproductions for situations for helping people out / having folks create small reproductions of their problems so we can help them better analyze their situation. Having "official" ember support would be beneficial so we don't get comments like this in Discord:

> "Do you know some site like stackblitz for ember ?"
> "I see that Ember is not support :( "

Issues that affect projects that have as many files as the default ember boilerplate need to be resolved:
 - https://github.com/stackblitz/webcontainer-core/issues/294
   - This happens *every time* you try saving with a server running

**Comment by EricSimons** [comment]
Oh I see, the start command is optimistic in opening the devserver. Definitely seems like the writes to the `dist` folder are the issue here. Is there a way to turn that feature off by chance? If not, we've also been talking about adding a `.stackblitzignore` or something akin that would solve this as well. Curious wyt?

**Comment by NullVoxPopuli** [comment]
I'd think stackblitz should respect .gitignore

**Comment by NullVoxPopuli** [comment]
@EricSimons ember now has dedicated repo / branch for Stackblitz: https://github.com/ember-cli/editor-output/tree/stackblitz-app-output

**Comment by NullVoxPopuli** [comment]
@jrvidal / @EricSimons - is there a place I can PR to add the link? or is the website private?

**Comment by EricSimons** [comment]
@NullVoxPopuli ah yeah our website is public, but I'll circle with the team tomorrow morning and get this updated on the homepage/dashboard shortly 🙏

**Comment by NullVoxPopuli** [comment]
what's the verdict? <3

**Comment by NullVoxPopuli** [comment]
@EricSimons how'd the circling go?

**Comment by sulco** [comment]
Hi @NullVoxPopuli! I'd be happy to make this happen, and we should be able to add the starter this week.

One concern I have atm is, since our free tier doesn't include binary files, and the starter includes the `construction.png` image, people who open it and are on the free tier – or not logged in – would now see the the "upgrade to pro account" message:
<img width="715" alt="image" src="https://user-images.githubusercontent.com/1511906/150787260-b906b0f2-3dba-4c3c-b4ea-6e48b6e8e7ef.png">

Do you think you could change it to an svg or a hotlinked image? Either should resolve the issue.

**Comment by NullVoxPopuli** [comment]
tldr: I think it's fine if we leave things as is -- the options for REPLs are a "best of no good options" sort of deal, and I think this can "be put up with"

We can't really swap out the default blueprint, as it's _nearly_ the same as what folks use on their local machines. And if they have dependencies installed locally (due to already making an ember app somewhere else), it's nice that they don't have to have internet to have the default welcome experience.



Questions though:
 - is there an option or something (in the .stackblitzrc) to make this modal a little less intrusive?
 - if binary files require pro account, how come it renders? (or maybe I don't understand?

**Comment by sulco** [comment]
That's too bad, it's not the best UX for a starter we'd want to put as an official one. I can't see any different options, but I'll discuss it with the team.

Regarding the questions:
- there is no option to change the modal
- what I believe happens is that when the project launches, the image goes through WebContainer – which works – and at the same time tries to persist it – which is not available

Where does the image come from btw? I wasn't able to locate it in the repository.

**Comment by NullVoxPopuli** [comment]
> Regarding the questions:

thanks for the answers!

> Where does the image come from btw? I

 - this dependency [ember-welcome-page](https://github.com/ember-cli/editor-output/blob/stackblitz-app-output/package.json#L50)
- provides this component: [`<WelcomePage />`](https://github.com/ember-cli/ember-welcome-page/blob/master/src/components/welcome-page.hbs#L4)
- which references (and packages) that image here: https://github.com/ember-cli/ember-welcome-page/tree/master/images
- which is declared as "to be merged" with the consuming app's build here: https://github.com/ember-cli/ember-welcome-page/blob/master/package.json#L119

(it's a common pattern in the ember-ecosystem to have libraries that may provide static assets (super handy for work projects, too))

> and at the same time tries to persist it 

can we opt out of that? I don't think we need that behavior?

> it's not the best UX for a starter we'd want to put as an official one.

I suppose if we mustn't have the image, I'd have to remove the welcome page entirely by removing the addon and adjusting the blueprint files. (in this auto-updating script that we use to update online-editor outputs): https://github.com/ember-cli/ember-cli/blob/master/dev/update-output-repos.js#L67
but then that leads to a less-pleasant experience for on-boarding / not seeing any links or how to get help :thinking:

**Comment by sulco** [comment]
Oh my, I've been digging deeper and it looks like that `dist` folder – or rather us lacking the possibility of ignoring it – is the ultimate culprit. 

I've just forked the repo and set the `"output-path"` to `"/.dist"` (dot-folders are ignored by default) and looks like this solves the issue. See https://stackblitz.com/fork/github/sulco/editor-output/tree/stackblitz-app-output .

I imagine this doesn't solve the issue as this would make the setup non-default thought, right?

**Comment by sulco** [comment]
Ok, I will try to push this paths-ignoring configuration to the team – can't promise any ETA, but I realize it would be super useful

**Comment by NullVoxPopuli** [comment]
> 've been digging deeper and it looks like that dist folder – or rather us lacking the possibility of ignoring it – is the ultimate culprit.

taking care of that would also fix: https://github.com/stackblitz/webcontainer-core/issues/294
(and potentially every other issue I've had with _using_ stackblitz)

> I imagine this doesn't solve the issue as this would make the setup non-default thought, right?

as far as output goes, I think that's _less_ important. 

But, anything against ignoring what's specified in .gitignore?

> I've just forked the repo and set the "output-path" to "/.dist"

Nice! I didn't actually know about that option in the `.ember-cli` file :upside_down_face: 

> Ok, I will try to push this paths-ignoring configuration to the team – can't promise any ETA, but I realize it would be super useful

Thanks for investigating!! <3

**Comment by sulco** [comment]
Good news! looks like we are going with `.gitignore`! Should also be sooner rather than later, probably before the end of the next week 🤞

**Comment by NullVoxPopuli** [comment]
yay!!! :tada: :taco: :partying_face:

**Comment by NullVoxPopuli** [comment]
success? @sulco <3

**Comment by sulco** [comment]
WIP! :)

**Comment by NullVoxPopuli** [comment]
success? @sulco <3

**Comment by MichalBryxi** [comment]
Right now the URL from the issue description (https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output?file=package.json) gives me:

```sh
~/projects/mklizrgop.github
❯ npm install && npm start
warn preInstall No repository field
┌ [1/3] 🔍  Resolving dependencies
└ Completed in 1.221s
┌ [2/3] 🚚  Fetching dependencies
│ info pruneDeps Excluding 2 dependencies. For more information use `--verbose`.
└ Completed in 3.674s
┌ [3/3] 🔗  Linking dependencies
└ Completed in 1.982s
info security We found `install` scripts which turbo skips for security reasons. For more information see
https://turbo.sh/install-scripts.
└─ core-js@2.6.12

success Saved lockfile "package-lock.json"
success Updated "package.json"

success Install finished in 7.626s
$ ember serve
TypeError: Failed to fetch
    at _0x530128.performHTTPCall (https://mklizrgop--github.w.staticblitz.com/blitz.05388ae171e3a0a12539f1efbda0fa1189d4f549.js:15:282948)

~/projects/mklizrgop.github 13s
❯ 
```

**Comment by NullVoxPopuli** [comment]
Today, I get this error:
```
success Updated "package.json"

success Install finished in 7.734s
$ ember serve
TypeError: Right-hand side of 'instanceof' is not an object
    at _0x585795.[Symbol.hasInstance] (https://qgnixoogb--github.w.staticblitz.com/blitz.5bca79c80956550a0ee18f24bf5b874e18bb8626.js:6:285951)
    at processPromiseRejections (https://qgnixoogb--github.w.staticblitz.com/blitz.5bca79c80956550a0ee18f24bf5b874e18bb8626.js:6:1008798)
    at processTicksAndRejections (https://qgnixoogb--github.w.staticblitz.com/blitz.5bca79c80956550a0ee18f24bf5b874e18bb8626.js:6:719888)
    at _0x4460f7 (https://qgnixoogb--github.w.staticblitz.com/blitz.5bca79c80956550a0ee18f24bf5b874e18bb8626.js:8:208957)
```

**Comment by NullVoxPopuli** [comment]
Now I have:
```
ReferenceError: Object is not defined
    at Function.build (/home/projects/github-k7htnb/node_modules/console-ui/node_modules/chalk/index.js:129:2)
    at Function.get (/home/projects/github-k7htnb/node_modules/console-ui/node_modules/chalk/index.js:61:17)
    ```

**Comment by fvsch** [comment]
If/when we end up being compatible with Ember.js, there is some good info about adding a “starter” project for Ember on stackblitz.com at: https://github.com/stackblitz/core/issues/1572

**Comment by NullVoxPopuli** [comment]
Looks like ember doesn't even use this esm package directly. Broccoli does: https://github.com/broccolijs/broccoli/issues/498

**Comment by NullVoxPopuli** [comment]
@EricSimons and @sulco I've made some progress with patch-package.

https://stackblitz.com/github/NullVoxPopuli/ember-cli-editor-output/tree/stackblitz-app-patched-broccoli-and-no-ember-data?file=README.md
and opened the following issues / PRs:
- https://github.com/emberjs/data/issues/8021
- https://github.com/broccolijs/broccoli/pull/499

Atm, I have two situations depending on.. maybe luck? maybe browser?

In firefox, I get this error:
```
$ ember serve
[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "TypeError: NetworkError when attempting to fetch resource.".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```

and in chrome I get this error:
```
$ ember serve
TypeError: Failed to fetch
    at _0x425d0a._fetch (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:15:406182)
    at _0x425d0a.performHTTPCall (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:15:405754)
    at _0x425d0a._write (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:15:399835)
    at _0x425d0a.writeLatin1String (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:15:390491)
    at _0xba2ea3.writeLatin1String (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:15:210978)
    at https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:6:783438
    at writeGeneric (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:6:783719)
    at Socket._writeGeneric (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:6:237637)
    at Socket._write (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:6:237795)
    at doWrite (https://kkvingabo--github.w.staticblitz.com/blitz.bff91798029e1af88e01551d056a592b2882b9ac.js:6:1012713)
```

**Comment by NullVoxPopuli** [comment]
@EricSimons / @sulco any debugging tips?

**Comment by NullVoxPopuli** [comment]
Things seem to be looking smooth now, actually: https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output?file=README.md


can we get Ember added to the home page? <3 https://stackblitz.com/ 
or in the frontend tab? https://stackblitz.com/?starters=frontend

**Comment by Polve** [comment]
Is there some kind of tutorial on how to setup an ember project with webcontainers?

I would like to try it very much!

**Comment by NullVoxPopuli** [comment]
if you have a node environment, you'd do:
```bash
npx ember-cli new my-app
cd my-app
npm start
```

**Comment by betocantu93** [comment]
Is there a way to make livereload work? I see the console rebuilding, but not the demo.

**Comment by NullVoxPopuli** [comment]
I wonder if you need to fork it first? I have live reload working here: https://stackblitz.com/edit/github-qivg2a?file=app%2Fcomponents%2Fnav.hbs,app%2Frouter.js

**Comment by betocantu93** [comment]
@NullVoxPopuli uff, it worked! I had to sign in for it to work. NICE. 

i'll refactor ASAP codesandbox for this  from ember-eui demos

**Comment by NullVoxPopuli** [comment]
you can put them all on a github repo and use specially crafted URLs to load all your demos from github :tada: 

```
https://stackblitz.com/github/{user-name}/{repo-name}/tree/{branch-name}
```

This way, if something goofy is going on, you have multiple options for loading your example code :tada:  (stackblitz, github.dev, local, etc)

**Comment by betocantu93** [comment]
do you have an example of this @NullVoxPopuli ?

**Comment by NullVoxPopuli** [comment]
Yeah, the automatically updated URL: 

https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output

Updates every release

**Comment by NullVoxPopuli** [comment]
We now have both JS and TS generated templates:

JS: https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output
TS: https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output-typescript

New automation to keep them in sync: https://github.com/ember-cli/ember-cli/pull/10211

**Comment by NullVoxPopuli** [comment]
@Nemikolh any updates? <3

**Comment by NullVoxPopuli** [comment]
No worries, @Nemikolh , lemme catch you up:

> From what I could gather, it seems the main blocker is support for the esm package.

I think that was only for old copies of the ember generated app blueprint -- the solution was just to _avoid_ `esm`.

Ember updates its app templates every 6 weeks, so things can progressively get fixed / updated on these links over time:
JS: https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output
TS: https://stackblitz.com/github/ember-cli/editor-output/tree/stackblitz-app-output-typescript

Using these links
    - in firefox on windows, no issues
    - in firefox on linux, no issues
    - in chrome, on windows, no issues

**Comment by Nemikolh** [comment]
Hey @NullVoxPopuli, just a heads up. We are currently making improvements to our starter list UI and I'll make sure your templates will be included. I don't have an ETA for now, but be reassured that we haven't forgotten about Ember!

_(and thank you so much for your message, it was super useful :heart_eyes: )_

**Comment by d3lm** [comment]
Can you remind of the ticket that was filed for the `esm` package specifically, or do you happen to have a project at hand that doesn't work in WebContainers that we could use?

**Comment by NullVoxPopuli** [comment]
Yeah, totally - and super appreciate that mentality!

Here is them`esm` issue(s):
- https://github.com/stackblitz/webcontainer-core/issues/692
- https://github.com/stackblitz/webcontainer-core/issues/691

Reproduction blitz: https://stackblitz.com/edit/github-8qeupk?file=package.json

Looks like both GH issues have stable/non-updating blitz examples

**Comment by NullVoxPopuli** [comment]
@d3lm any updates you can share about adding Ember to the site / template / project launcher?

as for the `esm` issue, it isn't present in the blitz links -- it was for much older version of the ember starter project(s) 🎉

**Comment by sulco** [comment]
@NullVoxPopuli we're adding several new starters to the platform **this week** – Ember is going to be among them! ❤️

**Comment by NullVoxPopuli** [comment]
oooo very exciting!!! 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉  🥳

**Comment by d3lm** [comment]
@NullVoxPopuli That's fantastic news! So it doesn't use the `esm` package anymore?

**Comment by NullVoxPopuli** [comment]
@d3lm kinda -- some legacy code paths still use it

```
❯ npx ember-cli@latest new my-app
# wait forever because npm is slow 🙃
❯ npm why esm
esm@3.2.25 dev
node_modules/esm
  esm@"^3.2.4" from broccoli@3.5.2
  node_modules/broccoli
    broccoli@"^3.5.2" from ember-cli@4.12.1
    node_modules/ember-cli
      dev ember-cli@"~4.12.1" from the root project
      peer ember-cli@"^3.2.0 || ^4.0.0" from ember-cli-dependency-checker@3.3.1
      node_modules/ember-cli-dependency-checker
        dev ember-cli-dependency-checker@"^3.3.1" from the root project
```

but (I haven't looked at how `esm` is used in these packages) this either doesn't affect anything when doing `npm start` or the issue has been resolved.
